rclpy/rclpy/clock.py

```diff
@@ -172,6 +174,68 @@ def callback_shim(jump_dict):
             clock=self, threshold=threshold, pre_callback=pre_callback,
             post_callback=post_callback)
 
+    def sleep_until(self, until: Time, context=None) -> bool:
+        """
+        Sleep until a Time on this Clock is reached.
+
+        When using a ROSClock, this may sleep forever if the TimeSource is misconfigured and the
+        context is never shut down.
+        ROS time being activated or deactivated causes this function to cease sleeping and return
+        False.
+
+        :param until: Time at which this function should stop sleeping.
+        :param context: Context which when shut down will cause this sleep to wake early.
+            If context is None, then the default context is used.
+        :return: True if until was reached, or False if it woke for another reason.
+        :raises ValueError: until is specified for a different type of clock than this one.
+        :raises NotInitializedException: context has not been initialized or is shutdown.
+        """
+        if context is None:
+            context = get_default_context()
+
+        if not context.ok():
+            raise NotInitializedException()
+
+        if until.clock_type != self._clock_type:
+            raise ValueError("until's clock type does not match this clock's type")
+
+        event = _rclpy.Event()
+        time_source_changed = False
+
+        def on_time_jump(time_jump: TimeJump):
+            """Wake when time jumps and is past target time."""
+            nonlocal time_source_changed
+
+            # ROS time being activated or deactivated changes the epoch, so sleep
+            # time loses its meaning
+            time_source_changed = (
+                time_source_changed or
+                ClockChange.ROS_TIME_ACTIVATED == time_jump.clock_change or
+                ClockChange.ROS_TIME_DEACTIVATED == time_jump.clock_change)
+
+            if time_source_changed or self.now() >= until:
+                event.set()
```

Comment:
why is the event triggered when the time source changes?
shouldn't still be checking it `now >= until` with the new time source?

Reply:
I don't think it should because `until` is specified with the old time source. It's possible for ROS time to be enabled and then `/clock` to publish really small numbers that would cause sleep to not wake for years. This might happen when playing an old bag file, or connecting to a simulator where time may be close to zero.

Reply:
Sounds good.
Is this consistent with rclpcpp behavior?

Reply:
>  Is this consistent with rclpcpp behavior?

Yup, rclcpp also returns `false` when the time source changes https://github.com/ros2/rclcpp/blob/b2b676d3172ada509e58fa552a676a446809d83c/rclcpp/test/rclcpp/test_time.cpp#L571-L585